modules/openapi-generator/src/main/resources/python/model_test.mustache

```diff
@@ -22,11 +24,20 @@ class Test{{classname}}(unittest.TestCase):
     def tearDown(self):
         pass
 
-    def test{{classname}}(self):
+    def make_instance(self, include_optional):
         """Test {{classname}}"""
-        # FIXME: construct object with mandatory attributes with example values
         # model = {{packageName}}.models.{{classFilename}}.{{classname}}()  # noqa: E501
-        pass
+        if (include_optional) :
+            return {{classname}}({{#vars}}
+                {{name}} = {{#example}}{{{.}}}{{/example}}{{^example}}None{{/example}}{{#hasMore}}, {{/hasMore}}{{/vars}})  # noqa: E501
+        else:
+            return {{classname}}({{#requiredVars}}
+                {{name}} = {{#example}}{{{.}}}{{/example}}{{^example}}None{{/example}}{{#hasMore}}, {{/hasMore}}{{/requiredVars}})  # noqa: E501
+
+    def test{{classname}}(self):
+        """Test {{classname}}"""
+        #inst_req_only = self.make_instance(include_optional=False)
```

Comment:
Why is this commented out?
If it is because of a discriminator issue, can you handle that issue in the Java code?

Reply:
It is exactly a problem with Discriminators, and solving it from the java code means being quite intrusive, having impact on models and apis.

The signature of Cat's constructor is:
`def __init__(self, declawed=None, local_vars_configuration=None)`
thus it cannot accept any `class_name`.
Anyway, if you look into test_cat, if I loop over #vars I have:
```
           return Cat(
                declawed = True)  # noqa: E501

```which is good, but by looping over #requiredVars I have:
```
            return Cat(
                class_name = '0')  # noqa: E501

```Thus, I have class_name - which should not be there, but is the property controlling the Discriminator - in the requiredVars but NOT in the vars.

Reply:
Let me look into this more. I don't understand why we can't loop over vars. Isn't that what our model template is doing? Is it that animal class tests are missing the discriminator?

Reply:
I isolated the issue. I put this code into the model*.moustache templates:
```
{{classname}}
{{classVarName}}
{{#vars}}
vars {{{name}}}
{{/vars}}

{{#allVars}}
allVars {{{name}}}
{{/allVars}}

{{#requiredVars}}
requiredVars {{{name}}}
{{/requiredVars}}

{{#optionalVars}}
optionalVars {{{name}}}
{{/optionalVars}}

{{#readOnlyVars}}
readOnlyVars {{{name}}}
{{/readOnlyVars}}

{{#readWriteVars}}
readWriteVars {{{name}}}
{{/readWriteVars}}

{{#parentVars}}
parentVars {{{.}}}
{{/parentVars}}

{{#mandatory}}
mandatory {{{.}}}
{{/mandatory}}

{{#allMandatory}}
allMandatory {{{.}}}
{{/allMandatory}}

```
Here is the result for Animal and Cat:
```
Animal
animal
vars class_name
vars color

allVars class_name
allVars color

requiredVars class_name

optionalVars color


readWriteVars class_name
readWriteVars color


mandatory class_name

allMandatory class_name

```

```
Cat
cat
vars declawed

allVars class_name
allVars color
allVars declawed

requiredVars class_name

optionalVars declawed
optionalVars color


readWriteVars declawed
readWriteVars class_name
readWriteVars color



allMandatory class_name

```

Here is the __init__ of Animal:
`    def __init__(self, class_name=None, color='red', local_vars_configuration=None):  # noqa: E501`
and finally the __init__ of Cat:
`    def __init__(self, declawed=None, local_vars_configuration=None):  # noqa: E501`

This, Animal is fine, but Cat should loop over allVars instead of vars when generating the __init__ function.

Do you agree with my diagnosis?

Reply:
Great job investigating. Wow, I didn't know about the mandatory and allMandatory properties, good find. Looking at the model mustache file, I disagree with your suggestion.
In our model mustache file here: https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/python/model.mustache#L60
we only loop over vars in the class `__init__` input signature.
So our solution here should only use vars or a subset of vars.

It's possible to only use var properties that are required using the #required property
defined here: https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenProperty.java#L55
We are currently using it in our template here:
https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/python/model.mustache#L72

How about changing your else clause in make_instance to only include the required vars?
```
        else:
            return {{classname}}({{#vars}}{{#required}}
                {{name}} = {{#example}}{{{.}}}{{/example}}{{^example}}None{{/example}}{{#hasMore}}, {{/hasMore}}{{/required}}{{/vars}})  # noqa: E501
```

Reply:
Done. I am still thinking that `required_vars` should be a subset of `vars`, but I don't know what is causing that.

Reply:
"that" = the discrepancy between `vars` and `required_vars`. I will investigate in the weekend if other generators have the same structures.